packages/cli-main/src/cli/services/upgrade.ts

```diff
@@ -0,0 +1,174 @@
+import {error, file, os, output, path, system} from '@shopify/cli-kit'
+import {
+  addNPMDependencies,
+  findUpAndReadPackageJson,
+  checkForNewVersion,
+  DependencyType,
+  getPackageManager,
+} from '@shopify/cli-kit/node/node-package-manager'
+
+// Canonical list of oclif plugins that should be installed globally
+const globalPlugins = ['@shopify/theme']
+
+interface PackageJsonContents {
+  name: string
+  dependencies?: {[name: string]: string}
+  devDependencies?: {[name: string]: string}
+  oclif?: {
+    plugins?: string[]
+  }
+}
+
+export async function upgrade(directory: string, currentVersion: string): Promise<void> {
+  let newestVersion: string | void
+
+  const projectDir = await getProjectDir(directory)
+  if (projectDir) {
+    newestVersion = await upgradeLocalShopify(projectDir, currentVersion)
+  } else if (usingPackageManager()) {
+    throw new error.Abort(
+      output.content`Couldn't find the configuration file for ${output.token.path(
+        directory,
+      )}, are you in a Shopify project directory?`,
+    )
+  } else {
+    newestVersion = await upgradeGlobalShopify(currentVersion)
+  }
+
+  if (newestVersion) {
+    output.success(`Upgraded Shopify CLI to version ${newestVersion}`)
+  }
+}
+
+async function getProjectDir(directory: string) {
+  return path.findUp(
+    async (dir: string) => {
+      const configFilesExist = await Promise.all(
+        ['shopify.app.toml', 'hydrogen.config.js', 'hydrogen.config.ts'].map(async (configFile) => {
+          return file.exists(path.join(dir, configFile))
+        }),
+      )
+      if (configFilesExist.some((bool) => bool)) return dir
+    },
+    {
+      cwd: directory,
+      type: 'directory',
+    },
+  )
+}
+
+async function upgradeLocalShopify(projectDir: string, currentVersion: string): Promise<string | void> {
+  const packageJson = (await findUpAndReadPackageJson(projectDir)).content as PackageJsonContents
+  const packageJsonDependencies = packageJson.dependencies || {}
+  const packageJsonDevDependencies = packageJson.devDependencies || {}
+
+  let resolvedVersion: string = {...packageJsonDependencies, ...packageJsonDevDependencies}[await cliDependency()]!
+  if (resolvedVersion.slice(0, 1).match(/[\^~]/)) resolvedVersion = currentVersion
+  const newestVersion = await checkForNewVersion(await cliDependency(), resolvedVersion)
+
+  if (!newestVersion) return wontInstall(resolvedVersion)
+
+  outputUpgradeMessage(resolvedVersion, newestVersion)
+
+  await installJsonDependencies('prod', packageJsonDependencies, projectDir)
```

Comment:
If we pass `packageJsonDependencies` aren't we updating all the dependencies existing in the project?

Reply:
No, since `installJsonDependencies filters to just `@shopify/cli` and canonical oclif plugins.